Issue #8891: Added a check in cse_main for matrices.

mihir-w opened the pull request:
`cse_main` was taking matrices as one whole unit (instead of breaking them down) if they occurred in lists. So I added a snippet in `cse_main` which checks whether any sub-expression is a matrix and then accordingly breaks it, finally appending it to `exprs`.

Edit: Changed postprocessing.

Fixes #8891 
@moorepants please review.

mihir-w committed:
Issue #8891: Added a check in cse_main for matrices.

mihir-w committed:
Issue #8891: Added postprocessing for correct canonical form with lists

smichr commented:
You might consider preserving the matrix types and watching out for SparseMatrix.

```
>>> m = eye(2)*(x+y)
>>> s = SparseMatrix.zeros(2);s[0,0] = x + y
>>> d = x + y, m, s
>>> p = [x + y, Tuple(*m._mat),
...             Tuple(*s._smat.items())]
>>> r,e = cse(p)
>>> c = []
>>> c.append(e[0])
>>> c.append(d[1].__class__(m.rows, m.cols, e[1]))
>>> c.append(d[2].__class__(s.rows, s.cols, {}))
>>> for k,v in e[2]:
...  c[-1][k] = v
...
>>> c
[x0, Matrix([
[x0,  0],
[ 0, x0]]), Matrix([
[x0, 0],
[ 0, 0]])]
>>> [i.__class__.__name__ for i in c]
['Symbol', 'MutableDenseMatrix', 'MutableSparseMatrix']
```

mihir-w commented:
Added the part regarding the SparseMatrix

> You might consider preserving the matrix types

Is this about the post-processing part where I am using `Matrix[....]` ?

Also, what are all the matrix types ( any links ) ?
I could only find the following: `sympy.matrices.dense.MutableDenseMatrix` and `sympy.matrices.immutable.ImmutableMatrix`. Any others ?

moorepants commented on sympy/simplify/cse_main.py:
```diff
@@ -458,6 +470,19 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
+        temp = []
+        i = 0
+        for e in exprs:
+            if isinstance(e, Matrix):
```
Here you should use the same class that was passed in.

mihir-w committed:
Issue #8891: Postprocessing done in same class

mihir-w commented:
Changed the line
```
temp.append([e.__class__(e.rows, e.cols, reduced_exprs[i:i+e.rows*e.cols])])
```

@moorepants does this seem correct ?

moorepants commented:
Yes, that seems right, but you should write a test for each matrix type.

Also, if you'd like to improve the docstring with some examples that would be awesome.

mihir-w commented:
Sure. Will do.

moorepants commented on sympy/simplify/tests/test_cse.py:
```diff
@@ -357,3 +357,9 @@ def test_issue_7840():
+    m1 = Matrix([w + x, w + x])
+    m2 = Matrix([y + z, y + z])
+    ans = '([(x0, w + x), (x1, y + z)], [[Matrix([\n[x0],\n[x0]])], [Matrix([\n[x1],\n[x1]])]])'
+    assert str(cse([m1, m2])) == ans
```
Doing this str() test is ok but not ideal because it misses checking anything else about the types. You should be testing to see if the objects are equivalent.

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -416,12 +416,23 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
+    copy = exprs
+    temp = []
+    for e in exprs:
+        if isinstance(e, Matrix) or isinstance(e, SparseMatrix):
```
cse should not treat SparseMatrix like a Matrix and generate potentially *many* zeros. Please see the code example that I gave where the pertinent data structures for each are used. Feel free to ask questions, too.

mihir-w replied:
So first I should do something like this:
```
temp.append(e.__class__(x.rows, x.cols, {}))
```
followed by filling the required spaces through the for loop you wrote
```
for k,v in e[2]:
...  c[-1][k] = v
```
right ?

mihir-w commented:
When I tried passing an `ImmutableMatrix` in `cse()` function, it raises a TypeError.
```
 raise TypeError("Data type not understood")
```
Is this normal ?

mihir-w committed:
Issue #8891: Used proper structures for sparse matrices, added tests and docstring

mihir-w commented:
Changed the code to use proper data structures for SparseMatrix.

Even if input is a single matrix, I have taken it as a list. It avoids duplication of code, as otherwise a check will be required to know whether `reduced_exprs` is a matrix or a list, followed by doing the same thing for each circumstance. 

Added docstring and improved tests.

smichr commented:
> Even if input is a single matrix, I have taken it as a list

It's ok to put it into a list for processing, but you should record at the start with a flag, 'is_matrix = True'. Then at the end before returning, check that flag and if true return `reduced_exprs[0]` (or something to that effect).

mihir-w committed:
Output is no longer lists for matrices. Changed few tests

mihir-w commented:
Matrices are now not returned as lists anymore. Changed an existing test in `test_cse.py` as it was based on list of matrix. Also changed the test for this issue.



> just set 0,0 (the other doesn't add anything to the example IMO).

No substitution would take place. An expression needs to occur at least twice before it can be substituted. Hence I let the part be. Added comma.

smichr committed:
cse: simplify docstring and restoration of matrices

smichr commented:
The [pr has been sent](https://github.com/mihir-w/sympy/pull/4)

smichr committed:
cse: handle immutable matrices

smichr commented:
The pr also handles the immutable matrix type.

mihir-w committed:
Merge pull request #4 from smichr/pr

cse: simplify docstring and restoration of matrices

smichr commented:
Sorry about the whitespace error. I found the mistake in the script that I use to remove it. I'll open a new PR and commit from there. Thanks! Folks that work with matrices will appreciate being able to use cse on them.
